content/client-lib-development-guide/features.textile

```diff
@@ -444,17 +453,28 @@ h3(#realtime-channel). Channel
+* @(RTP2)@ A "@PresenceMap@":https://github.com/ably/ably-java/blob/master/src/io/ably/realtime/Presence.java#L384-L529 should be used to maintain a list of members present on a channel. As there are no guarantees that during the @SYNC@ phase presence events will arrive in order i.e. a leave event for a member can arrive before that member is later registered as as present as part of the initial @SYNC@ operation. As such, until the @SYNC@ operation is complete, timestamps must be recorded for all members to ensure the most recent present state is used, see the "Java implementation":https://github.com/ably/ably-java/blob/master/src/io/ably/realtime/Presence.java#L418-L431.  Once a @SYNC@ event is complete, the members that are now considered @ABSENT@ can be removed from the map, see the "Java implementation":https://github.com/ably/ably-java/blob/master/src/io/ably/realtime/Presence.java#L504-L509
 * @(RTP3)@ If a @SYNC@ operation is underway but not yet complete, and the transport is disconnected unexpectedly, then if the connection is resumed successfully, it is the responsibility of the client library to complete the @SYNC@ operation. The client library requests a @SYNC@ resume by sending a @SYNC@ @ProtocolMessage@ with the last received sync serial number. See the "Ruby implementation":https://github.com/ably/ably-ruby/blob/7b18a20/lib/ably/realtime/presence/members_map.rb#L169-L176 and the "Ruby test":https://github.com/ably/ably-ruby/blob/7b18a20/spec/acceptance/realtime/presence_spec.rb#L1338-L1357
+* @(RTP18)@ The realtime system reserves the right to initiate a sync of the presence members at any point once a channel is attached and presence is in sync. The server initiated sync allows it to send a complete list of members present on the channel at that point in time. A sync is started by the server when it sends a @SYNC@ @ProtocolMessage@ to the client containing a @channel@ attribute <br>**TBC - how does the client lib know definitely it's starting a @SYNC@ and not continuing one, and should the client lib continue the @SYNC@ if disconnected & reconnected usingt the cursor?**.
```

Comment:
The way this works is that a cursor value sent by realtime (in the `channelSerial` field) is a two-part identifier `<sync sequence id>:<cursor value>`. The `sync sequence id` is a new id when a sync sequence starts.

There are 2 ways we could do this:

- say that the realtime system will only perform a sync after sending an `ATTACHED` message with a `sync` flag. This is certainly true now (except for syncs requested by the client).

- say that the library remembers the `sync sequence id`; that value is cleared on completion of any sync, and set when there is a sync message. This means that the library can detect an incomplete sync if it receives an unexpected sync message with the wrong sequence id.

If the transport drops, the client can continue a sync sequence that was in progress prior to the disconnection provided that the resume connection serial it provides ensures that it has already processed, or gets from the backlog, all presence messages on the channel from the point that the first sync message in the sequence was received. If the realtime system isn't happy continuing that sync sequence, it will start a new one, and the sync sequence identifier will allow the client to know that it's a new sequence.